File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        ret_val = 0
        if len(s) <= 1:
            return ret_val
        dict = {}
        for i in range(len(s)):
            if s[i] in dict:
                dict[s[i]].append(i)
            else:
                dict[s[i]] = [i]
        arr = []
        for j in dict.values():
            arr.append(len(j))
        while len(set(arr)) != len(arr):
            for k in range(1, len(arr)):
                if arr[k] == 0:
                    arr.pop(k)
                    break
                if arr[k] in arr[0:k]:
                    arr[k] += -1
                    ret_val += 1
                    break
        return ret_val

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.rb
# @param {String} s
# @return {Integer}
def min_deletions(s)
    output = 0
    return output if s.length <= 1
    hash = Hash.new { |h, k| h[k] = [] }
    s.each_char.with_index { |char, idx| hash[char] << idx }
    arr = []
    hash.values.each { |subarr| arr << subarr.length }
    until arr == arr.uniq
        (1...arr.length).each do |idx|
            if arr[idx] == 0
                arr.delete_at(idx)
                break
            end
            if arr[0...idx].include?(arr[idx])
                arr[idx] -= 1
                output += 1
                break
            end
        end
    end
    output
end